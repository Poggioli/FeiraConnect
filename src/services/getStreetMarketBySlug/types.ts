import { UndefinedInitialDataOptions } from "@tanstack/react-query";
import { AxiosError } from "axios";
import { GetStreetMarketsByCity } from "@/services/getStreetMarketsByCity";

export type UseGetStreetMarketBySlug = GetStreetMarketsByCity

export type UseGetStreetMarketBySlugQueryKey = ["getStreetMarketBySlug", string];

export type UseGetStreetMarketBySlugOptions = UndefinedInitialDataOptions<
  UseGetStreetMarketBySlug,
  AxiosError,
  UseGetStreetMarketBySlug,
  UseGetStreetMarketBySlugQueryKey
>
