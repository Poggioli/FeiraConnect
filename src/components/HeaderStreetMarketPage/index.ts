export * from "./HeaderStreetMarketPage";
export * from "./useHeaderStreetMarketPage";