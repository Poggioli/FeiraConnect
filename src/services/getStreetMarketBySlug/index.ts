export * from "./getStreetMarketBySlug";
export * from "./types";