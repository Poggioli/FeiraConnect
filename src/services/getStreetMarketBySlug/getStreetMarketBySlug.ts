import { api } from "@/services/api";
import { useQuery } from "@tanstack/react-query";
import { AxiosError } from "axios";
import {
  UseGetStreetMarketBySlug, UseGetStreetMarketBySlugOptions, UseGetStreetMarketBySlugQueryKey
} from "./types";

const queryKey = "getStreetMarketBySlug" as const;
const url = 'street-markets' as const;

export function createGetStreetMarketBySlugQueryKey(
  slug: string
): UseGetStreetMarketBySlugQueryKey {
  return [queryKey, slug];
}

export function useGetStreetMarketBySlug(options: UseGetStreetMarketBySlugOptions) {
  return useQuery<
    UseGetStreetMarketBySlug,
    AxiosError,
    UseGetStreetMarketBySlug,
    UseGetStreetMarketBySlugQueryKey
  >({
    ...options,
    queryKey: options.queryKey,
    queryFn: async ({ queryKey: { "1": slug } }) => {
      const res = await api.get<UseGetStreetMarketBySlug>(`${url}/${slug}`);
      return res.data;
    }
  })
}