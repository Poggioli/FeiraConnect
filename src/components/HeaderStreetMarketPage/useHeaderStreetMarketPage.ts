import { createGetStreetMarketBySlugQueryKey, useGetStreetMarketBySlug } from "@/services/getStreetMarketBySlug";
import { useParams } from "@tanstack/react-router";

export function useHeaderStreetMarketPage() {
  const { streetMarket } = useParams({ from: "/city/_$city/$streetMarket" });
  const streetMarketService = useGetStreetMarketBySlug({
    queryKey: createGetStreetMarketBySlugQueryKey(streetMarket),
  });

  return {
    streetMarketService,
  };
}
