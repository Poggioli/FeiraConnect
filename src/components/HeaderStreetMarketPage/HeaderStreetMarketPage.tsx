import { FC } from "react";
import { useHeaderStreetMarketPage } from "./useHeaderStreetMarketPage";
import { notFound } from "@tanstack/react-router";
import { Skeleton } from "@/components/ui/skeleton";
import { TitlePage } from "@/components/TitlePage";
import { Badge } from "../ui/badge";
import { MapPinned } from "lucide-react";
import { Helmet } from "react-helmet";

export const HeaderStreetMarketPage: FC = () => {

  const {
    streetMarketService: {
      isFetching,
      error,
      data,
    }
  } = useHeaderStreetMarketPage();

  if (error && error.response?.status === 404) {
    throw notFound({ routeId: '/city/_$city/$streetMarket' })
  }

  return isFetching ? (
    <>
      <Skeleton className="min-h-10 w-32" />
      <Skeleton className="min-h-5 w-16" />
    </>
  ) : (
    <>
      <Helmet>
        <title>{data?.name} | Feira connect</title>
      </Helmet>
      <TitlePage>{data?.name}</TitlePage>
      <a
        href={data?.location}
        target="_blank"
        className="leading-[initial]"
        aria-label={`Localização da feira ${data?.name}`}
      >
        <Badge className="gap-3">
          local
          <MapPinned className="h-4 w-4" />
        </Badge>
      </a>
    </>
  );
};
