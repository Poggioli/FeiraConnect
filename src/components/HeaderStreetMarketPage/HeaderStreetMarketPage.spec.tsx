import { waitFor } from "@testing-library/dom";
import { customRender } from "@tests/customRender";
import { getStreetMarketBySlugHandlers } from "@tests/getStreetMarketBySlug.mock";
import { server } from "@tests/server.mock";
import { HeaderStreetMarketPage } from "./HeaderStreetMarketPage";

const mockUseParams = vi.fn();
const mockNotFound = vi.fn();

vi.mock("@tanstack/react-router", async (importOriginal) => {
  const original = await importOriginal<Record<string, unknown>>();
  return {
    ...original,
    useParams: (args: unknown) => mockUseParams(args),
    notFound: (args: unknown) => mockNotFound(args)
  };
});

describe("@/components/HeaderStreetMarketPage", () => {
  beforeEach(() => {
    mockUseParams.mockReturnValue({ streetMarket: "streetMarket" });
  });

  it(`GIVEN a HeaderStreetMarketPage
      WHEN server return success
      THEN should show the name of city`, async () => {
    server.use(getStreetMarketBySlugHandlers.filled);
    const { getByText } = customRender(<HeaderStreetMarketPage />);

    await waitFor(() => {
      const title = getByText("name 1");
      const local = getByText("local");
      expect(local).toBeVisible();
      expect(local.parentElement).toHaveAttribute("href")
      expect(title).toBeVisible();
    })
  });

  it(`GIVEN a HeaderStreetMarketPage
      WHEN not finish load
      THEN should show the loading state`, () => {
    server.use(getStreetMarketBySlugHandlers.filledWithDelay);
    const { baseElement } = customRender(<HeaderStreetMarketPage />);

    expect(baseElement).toMatchSnapshot();
  });

  it(`GIVEN a HeaderStreetMarketPage
      WHEN response is 404
      THEN should call notFound method from @tanstack/react-router`, async () => {
    server.use(getStreetMarketBySlugHandlers.error);
    const { getByText } = customRender(<HeaderStreetMarketPage />);

    await waitFor(() => {
      const errorMessage = getByText('Error Boundary')
      expect(errorMessage).toBeVisible();
      expect(mockNotFound).toHaveBeenCalled();
      expect(mockNotFound).toHaveBeenCalledWith({ routeId: "/city/_$city/$streetMarket" });
    })
  });
});
