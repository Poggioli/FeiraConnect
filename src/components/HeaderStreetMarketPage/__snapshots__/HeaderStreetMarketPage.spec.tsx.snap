// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@/components/HeaderStreetMarketPage > GIVEN a HeaderStreetMarketPage
      WHEN not finish load
      THEN should show the loading state 1`] = `
<body>
  <div />
</body>
`;
